### ISSUE TITLE
Bailout called.  Further testing stopped:  *E* cached folder is missing

### ISSUE THREAD

**Issue by szabgab** [jgpuckering/Weather-GHCN-Fetch#1]
While trying to run the tests on a clean environment I get:
```
...
t/21_station.t .................. ok
t/22_station_table_subs.t ....... ok
Bailout called.  Further testing stopped:  *E* cached folder is missing
FAILED--Further testing stopped: *E* cached folder is missing
make: *** [Makefile:966: test_dynamic] Error 255
error running make test
Error: Process completed with exit code 2.
```

**Comment by jgpuckering** [comment]
Hi Gabor,

This is odd.  I'm not seeing this problem in any of the other tests that ran last night.  It obviously arises from this bit of code in the test script:

my $Bin = $FindBin::Bin;
my $Cachedir = path($Bin,'ghcn_cache')->stringify;

if (not -d $Cachedir) {
    BAIL_OUT "*E* cached folder is missing";
}

The folder t\ghcn_cache should have been created when the distro was unpacked.  I put this bit of code in as a check, just in case something bizarre happened.

This version of Weather::GHCN::Fetch passed testing on 82 platforms last night, failing in 5 -- but not for the reason you've reported in your CI test.  Those 5 failures were all in 91_ghcn_fetch.t when testing -help.  Looks like a multibyte string returned by POD =header1.

On my system, when I run prove t\22_station_table_subs.t, it CD's to the t folder and FindBin generates an absolute path to it.  The code above concatenates that path with ghcn_cache, and the path exists because it was unpacked from the tarball.

Something different seems to be happening in the CI test environment - but I don't know how to reproduce or investigate it.

Either t\ghcn_cache isn't there (which is doubtful), or the execution folder for the test run is not the t folder.

If you could verify these two things it would be a big help:

  1.  When your test machinery unpacks the Weather::GHCN::Fetch.tar.gz file, is there a t\ghcn_cache folder?
  2.  When a test executes on your platform, what path does FindBin::Bin return?  The t folder?  The parent of t (where prove usually runs)?  Or something else?

As a final note, since this is the first time we've corresponded, my thanks to you for all the work you've done over the years on the perl platform and with your blog posts.  I've been the beneficiary of your work many times.  And if I had a way to figure this out without your help, I wouldn't bother you with it.

Regard,

Gary Puckering


From: Gabor Szabo ***@***.***>
Sent: October 17, 2022 2:03 AM
To: jgpuckering/Weather-GHCN-Fetch ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [jgpuckering/Weather-GHCN-Fetch] Bailout called. Further testing stopped: *E* cached folder is missing (Issue #1)


While trying to run the tests on a clean environment I get:

...

t/21_station.t .................. ok

t/22_station_table_subs.t ....... ok

Bailout called.  Further testing stopped:  *E* cached folder is missing

FAILED--Further testing stopped: *E* cached folder is missing

make: *** [Makefile:966: test_dynamic] Error 255

error running make test

Error: Process completed with exit code 2.

-

**Comment by szabgab** [comment]
hi Gary,
I saw you already merged the PR with the CI configuration. 
It uses the files in the repository, not the ones from the released version that is the .tag.gz file.
I assume there is some script that creates and populate `t\ghcn_cache` before you create the tar.gz file. How is it done?

**Comment by szabgab** [comment]
I tried using the `bin/ghcn_fetch` script, but failed:

```
perl -Ilib bin/ghcn_fetch 
*E* cache directory does not exist at lib/Weather/GHCN/StationTable.pm line 1319.
```

**Comment by jgpuckering** [comment]
If t\ghcn_cache is empty, then executing the test scripts (I.e. running prove) will cause it to be populated with the same files that I’ve provided in the .tag.gz file – well, possibly a more up to date version of them.  The tests have been designed to run to be insensitive to the date of the cached files.

I prepopulate ghcn_cache in the distro so that the tests can execute without an internet connection.  Also, they run faster since they don’t have to download those files.


Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by szabgab** [comment]
I created the ghcn_cache locally, but I still get the error message. If the tests need that folder, then  I guess the tests should create it, if it does not exist yet.

**Comment by jgpuckering** [comment]
Weird.  The tarbal contains that folder and it’s content.  It lives within the t folder, beneath the root of the tarball.  Or at least, it should.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by szabgab** [comment]
you might try a clean git clone o n your computer to reproduce the situation.

**Comment by jgpuckering** [comment]
The path being reported by the latest test that is bailing out is:

/home/runner/work/Weather-GHCN-Fetch/Weather-GHCN-Fetch/.build/0r8ga6_eAF/t/ghcn_cache

Now, if the tarball was extracted into /home/runner/work/Weather-GHCN-Fetch and then dzil test is run, then that would create a .build folder and I guess the temporary folder 0r8ga6_eAF.  The “prove” command should be running from within that folder, beneath which is t.  Beneath that is ghcn_cache.

I think what you are telling me, though, is that ghcn_cache doesn’t exist in this location.  But I don’t understand why it wouldn’t be.  It should just be created during the tarball extraction.

This doesn’t seem to happen in the CPAN test environment, or in mine.  FindBin finds the test script in the t folder, and then I use that to construct an absolute path to t\ghcn_cache.

I’m mystified.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by jgpuckering** [comment]
Damn!  I tried a git clone as you suggested.  There’s no ghcn_cache folder.

Ok, well I don’t understand how it worked in the CPAN tests but I’ll re-add ghcn_cache and push it to the remote repo.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by szabgab** [comment]
There is no tarball in CI. It is running from the cloned git repository. A clean clone of he repo.

**Comment by jgpuckering** [comment]
Yeah, I get it now.  And ghcn_cache was never added to the repo!  Honestly, I don’t know how I missed that.  I thought it had been added and committed long ago.  Sorry for the work that caused you.

I’ve added, committed and pushed it to GitHub – and verified it is there.

Hopefully now the CI tests will work!  Turns out the message wasn’t erroneous after all.  The folder really was missing!

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by szabgab** [comment]
The tests now pass on my local computer, but there are still some issues in the CI environment.

While adding the cached files to the repo works now, I am not sure if this is the best approach. 

I just noticed something. If I clean the t/ghcn_cache/ directory and then

```
cd t
prove .
```
then the tests pass (and the t/ghcn_cache/ directory is populated), but if I am in the root of the project and I run
`dzil test` then I get the error as earlier.

I think the usual way is to run the tests while the current working directory is the root of the project and that the test should all run regardless of the current working directory.

Oh and BTW there is no need to apologize. Part of the  reason to have a CI is to figure these things out and make it easy to everyone to contribute to the project

**Comment by jgpuckering** [comment]
In my experience, prove works more reliably if run in the folder above t.  But maybe there’s a better way to set up test scripts than the way I do it.

Incidentally, 00_initialize_test_env.t will clean the cache if give a “clean” option on the command line.

It’s getting pretty late in my timezone.  I’ll do some experimenting with prove tomorrow and then reconsider the way my tests get the paths they need.

At least tonight has gotten me past the two main problems that were happening!

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by jgpuckering** [comment]
This issue is verified as fixed in v0.0.008.
